pmd-java/src/main/java/net/sourceforge/pmd/lang/java/oom/rule/CyclomaticComplexityRule.java

```diff
@@ -0,0 +1,118 @@
+        "reportLevel", "Cyclomatic Complexity reporting threshold", 1, 30, 10, 1.0f);
+
+    public static final BooleanProperty SHOW_CLASSES_COMPLEXITY_DESCRIPTOR = new BooleanProperty(
+        "showClassesComplexity", "Add class average violations to the report", true, 2.0f);
+
+    public static final BooleanProperty SHOW_METHODS_COMPLEXITY_DESCRIPTOR = new BooleanProperty(
+        "showMethodsComplexity", "Add method average violations to the report", true, 3.0f);
+
+
+    private static final String[] VERSION_LABELS = {"standard", "ignoreBooleanPaths"};
+
+    private static final MetricVersion[] CYCLO_VERSIONS = {Metric.Version.STANDARD, CycloMetric.Version.IGNORE_BOOLEAN_PATHS};
+
+    public static final EnumeratedProperty<MetricVersion> CYCLO_VERSION_DESCRIPTOR = new EnumeratedProperty<>(
+        "cycloVersion", "Choose a variant of Cyclo or the standard",
+        VERSION_LABELS, CYCLO_VERSIONS, 0, 3.0f);
+
+    private int reportLevel;
+    private boolean showClassesComplexity = true;
+    private boolean showMethodsComplexity = true;
+    private MetricVersion cycloVersion = Version.STANDARD;
+
+
+    public CyclomaticComplexityRule() {
+        definePropertyDescriptor(REPORT_LEVEL_DESCRIPTOR);
+        definePropertyDescriptor(SHOW_CLASSES_COMPLEXITY_DESCRIPTOR);
+        definePropertyDescriptor(SHOW_METHODS_COMPLEXITY_DESCRIPTOR);
+        definePropertyDescriptor(CYCLO_VERSION_DESCRIPTOR);
+    }
+
+
+    @Override
+    public Object visit(ASTCompilationUnit node, Object data) {
+        reportLevel = getProperty(REPORT_LEVEL_DESCRIPTOR);
+        showClassesComplexity = getProperty(SHOW_CLASSES_COMPLEXITY_DESCRIPTOR);
+        showMethodsComplexity = getProperty(SHOW_METHODS_COMPLEXITY_DESCRIPTOR);
+        Object version = getProperty(CYCLO_VERSION_DESCRIPTOR);
+        cycloVersion = version instanceof MetricVersion ? (MetricVersion) version : Version.STANDARD;
+
+        super.visit(node, data);
+        return data;
+    }
+
+    // TODO:cf consider enum classes too (eg create some ASTAnyTypeDeclaration umbrella interface)
+    @Override
+    public Object visit(ASTClassOrInterfaceDeclaration node, Object data) {
+        if (node.isInterface()) {
+            return data;
+        }
+
+        super.visit(node, data);
+        if (showClassesComplexity) {
+            int classCyclo = (int) Metrics.get(ClassMetricKey.CYCLO, node, cycloVersion);
+            int classHighest = (int) Metrics.get(OperationMetricKey.CYCLO, node, cycloVersion, ResultOption.HIGHEST);
+
+            if (classCyclo >= reportLevel || classHighest >= reportLevel) {
+                addViolation(data, node,
+                             new String[] {"class", node.getImage(), classCyclo + " (Highest = " + classHighest + ')'});
+            }
+        }
+        return data;
+    }
+
+    // TODO:cf consider merging these two methods (changes to JavaParserVisitor)
```

Comment:
Hum ok. Maybe a simpler approach would be to use an adapter that defines `visit(ASTMethodOrConstructorDeclaration, Object)` and delegates to the concrete nodes.